provider/datagen/src/baked_exporter.rs

```diff
@@ -330,80 +334,95 @@ impl DataExporter for BakedExporter {
-                            Ok(Self::#singleton_ident)
-                        } else {
-                            Err(icu_provider::DataErrorKind::MissingLocale)
-                        }
-                    }
+            if locale != "und" {
+                Err(DataError::custom("Singleton mode needs the default locale"))?;
+            }
+
+            let singleton_ident = format!("SINGLETON_{}", ident.to_ascii_uppercase())
+                .parse::<TokenStream>()
+                .unwrap();
+
+            // Exposing singleton structs separately allows us to get rid of fallibility by using
+            // the struct directly.
+            singleton = Some(quote! {
+                #[clippy::msrv = "1.61"]
+                impl $provider {
+                    #[doc(hidden)]
+                    pub const #singleton_ident: &'static #struct_type = &#bake;
+                }
+            })
+            .into_iter();
+
+            quote! {
+                if locale.is_empty() {
+                    Ok(Self::#singleton_ident)
                 } else {
-                    quote! {
-                        if locale.strict_cmp(#locale.as_bytes()).is_eq() {
-                            Ok(Self::#singleton_ident)
-                        } else {
-                            Err(icu_provider::DataErrorKind::MissingLocale)
-                        }
-                    }
+                    Err(icu_provider::DataErrorKind::ExtraneousLocale)
                 }
             }
-            _ => {
-                let mut map = BTreeMap::new();
-                let mut statics = Vec::new();
-
-                for (bake, locales) in values {
-                    let first_locale = locales.iter().next().unwrap();
-                    let anchor = syn::parse_str::<syn::Ident>(
-                        &first_locale.to_ascii_uppercase().replace('-', "_"),
-                    )
-                    .unwrap();
-                    statics.push(quote! { static #anchor: #struct_type = #bake; });
-                    map.extend(locales.into_iter().map(|l| (l, anchor.clone())));
-                }
+        } else {
+            let mut map = BTreeMap::new();
+            let mut statics = Vec::new();
+
+            for (bake, locales) in values {
+                let first_locale = locales.iter().next().unwrap();
+                let anchor = syn::parse_str::<syn::Ident>(
+                    &first_locale.to_ascii_uppercase().replace('-', "_"),
+                )
+                .unwrap();
+                statics.push(quote! { static #anchor: #struct_type = #bake; });
+                map.extend(locales.into_iter().map(|l| (l, anchor.clone())));
+            }
+
+            let (keys, values): (Vec<_>, Vec<_>) = map.into_iter().unzip();
 
-                let (keys, values): (Vec<_>, Vec<_>) = map.into_iter().unzip();
+            let lookup = quote! {
```

Comment:
suggestion: might benefit in clarity by being an additional method/function since otherwise we have a giant block in an `if` predicate below

Reply:
I think we'd lose a lot of clarity here and don't find it's a big issue in the generated code.